File: app/src/main/java/com/nononsenseapps/feeder/ui/text/ImageTextLoader.java
package com.nononsenseapps.feeder.ui.text;

import android.content.Context;
import android.graphics.Point;
import android.support.v4.content.AsyncTaskLoader;
import android.text.Spanned;

public class ImageTextLoader extends AsyncTaskLoader<Spanned> {

    private final String text;
    private final Point maxSize;
    private final boolean allowDownload;

    public ImageTextLoader(Context context, String text, Point maxSize, boolean allowDownload) {
        super(context);
        this.text = text;
        this.maxSize = maxSize;
        this.allowDownload = allowDownload;
    }

    @Override
    public Spanned loadInBackground() {
        return HtmlConverter.toSpannedWithImages(text, getContext(), maxSize, allowDownload);
    }

    @Override
    protected void onStartLoading() {
        forceLoad();
    }

    @Override
    protected void onStopLoading() {
        cancelLoad();
    }

    @Override
    protected void onReset() {
        super.onReset();

        // Ensure the loader is stopped
        onStopLoading();
    }
}

File: app/src/main/java/com/nononsenseapps/feeder/ui/text/ImageTextLoader.kt
package com.nononsenseapps.feeder.ui.text

import android.content.Context
import android.graphics.Point
import android.support.v4.content.AsyncTaskLoader
import android.text.Spanned
import android.util.Log
import java.net.URL

class ImageTextLoader(context: Context,
                      private val text: String,
                      private val siteUrl: URL,
                      private val maxSize: Point,
                      private val allowDownload: Boolean) : AsyncTaskLoader<Spanned>(context) {

    override fun loadInBackground(): Spanned? =
            try {
                toSpannedWithImages(context, text, siteUrl, maxSize, allowDownload)
            } catch (t: Throwable) {
                // It has proven impossible to prevent android.database.StaleDataException errors (cursor closed)
                Log.e("ImageTextLoader", "Error during load: $t")
                null
            }

    override fun onStartLoading() {
        forceLoad()
    }

    override fun onStopLoading() {
        cancelLoad()
    }

    override fun onReset() {
        super.onReset()

        // Ensure the loader is stopped
        onStopLoading()
    }
}
